sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveTableScanExec.scala

```diff
@@ -0,0 +1,501 @@
+      val fs                 = path.getFileSystem(hadoopConf)
+      val dirContents        = fs.listStatus(path).filter(isNonEmptyDataFile)
+      val partitionDirectory = PartitionDirectory(partValues, dirContents)
+      val maxSplitBytes      = FilePartition.maxSplitBytes(sparkSession, Array(partitionDirectory))
+
+      val splitFiles: Seq[PartitionedFile] = dirContents.flatMap { f =>
+        PartitionedFileUtil.splitFiles(
+          sparkSession,
+          f,
+          f.getPath,
+          isSplitable = true,
+          maxSplitBytes,
+          partitionDirectory.values
+        )
+      }.sortBy(_.length)(implicitly[Ordering[Long]].reverse)
+      FilePartition.getFilePartitions(sparkSession, splitFiles, maxSplitBytes)
+    }
+
+    // TODO [future]: Handle small-file optimization.
+    //                Currently assuming per-file reading.
+    SparkShimImpl.getFileScanRDD(sparkSession, readFile, filePartitions, readSchema)
+  }
+
+  private def createReadRDDForTable(
+                readFile: PartitionedFile => Iterator[InternalRow],
+                hiveTableRelation: HiveTableRelation,
+                readSchema: StructType,
+                sparkSession: SparkSession,
+                hadoopConf: Configuration
+              ) = {
+    val tableLocation: URI = hiveTableRelation.tableMeta.storage.locationUri.getOrElse{
+      throw new UnsupportedOperationException("Table path not set.")
+    }
+    // No need to check if table directory exists.
+    // FileSystem.listStatus() handles this for GpuHiveTableScanExec,
+    // just like for Apache Spark.
+    createReadRDDForDirectories(readFile,
+                                Array((tableLocation, InternalRow.empty)),
+                                readSchema,
+                                sparkSession,
+                                hadoopConf)
+  }
+
+  private def createReadRDDForPartitions(
+                readFile: PartitionedFile => Iterator[InternalRow],
+                hiveTableRelation: HiveTableRelation,
+                readSchema: StructType,
+                sparkSession: SparkSession,
+                hadoopConf: Configuration
+              ): RDD[InternalRow] = {
+    val partitionColTypes = hiveTableRelation.partitionCols.map(_.dataType)
+    val dirsWithPartValues = prunedPartitions.map { p =>
+      // No need to check if partition directory exists.
+      // FileSystem.listStatus() handles this for GpuHiveTableScanExec,
+      // just like for Apache Spark.
+      val uri = p.getDataLocation.toUri
+      val partValues: Seq[Any] = {
+        p.getValues.asScala.zip(partitionColTypes).map {
+          case (value, dataType) => castFromString(value, dataType)
+        }
+      }
+      val partValuesAsInternalRow = InternalRow.fromSeq(partValues)
+
+      (uri, partValuesAsInternalRow)
+    }
+
+    createReadRDDForDirectories(readFile,
+      dirsWithPartValues, readSchema, sparkSession, hadoopConf)
+  }
+
+  lazy val inputRDD: RDD[InternalRow] = {
```

Comment:
nit: Could you add a comment about why this is an RDD[InternalRow] but we read it as an RDD[ColumnarBatch]?

Reply:
Hmm. I've fashioned this similarly to how `inputRDD` is used in [`GpuFileSourceScanExec`](https://github.com/NVIDIA/spark-rapids/blob/branch-22.12/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileSourceScanExec.scala#L335). Now that you mention it, I'm not sure I know why it's so. 

In both cases, it seems to amount to `SparkShimImpl.getFileScanRDD()` returning an `RDD[InternalRow]`.